### PR TITLE
[usm] system-probe, add Postgres message counters to the eBPF module

### PR DESCRIPTION
**Pull request by yuri-lipnesh**
### What does this PR do?
Adds Postgres message counters to the eBPF module of the system-probe

### Motivation
PostgreSQL communication is done via a continuous stream of messages. This code gathers statistics on the number of Postgres messages processed by eBPF, providing valuable telemetry that enables optimization of eBPF buffer sizes for handling Postgres messages in the agent and system-probe processes.

### Describe how to test/QA your changes
Passed unit tests for Postgres monitoring on local VM.
This branch will be tested on staging environment before merge.

### Possible Drawbacks / Trade-offs
The complexity of an eBPF program will increase.

### CODE REVIEW

File: pkg/network/protocols/postgres/telemetry.go [hunk 1]
```diff
@@ -178,3 +179,61 @@ func (t *Telemetry) Log() {
...
+
+	for i := range t.msgCountBuckets {
+		v := kernCounts.Messages_count_buckets[i]
+		t.msgCountBuckets[i].Set(int64(v), isTLS)
+	}
+	if isTLS {
+		t.msgCountTlsPrev = *kernCounts
+	} else {
+		t.msgCountPlainPrev = *kernCounts
+	}
```
Comment (guyarb) [inline pkg/network/protocols/postgres/telemetry.go:227]: where do we use those values?

File: pkg/network/usm/postgres_monitor_test.go [hunk 2]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+		utils.WaitForProgramsToBeTraced(t, "go-tls", os.Getpid(), utils.ManualTracingFallbackEnabled)
+	}
+	pgClient := setupPGClient(t, serverAddress, isTLS)
+	defer func() {
+		_ = pgClient.RunQuery(dropTableQuery)
+		if pgClient != nil {
+			pgClient.Close()
+			pgClient = nil
+		}
+	}()
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:940]: if `pgClient` == nil -> you'll panic when trying to call `RunQuery`

File: pkg/network/usm/postgres_monitor_test.go [hunk 3]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+
+	for i := 0; i < ebpf.MaxBucketsNotFragmented; i++ {
+		testName := fmt.Sprintf("kernel messages count bucket[%d]", i)
+		t.Run(testName, func(t *testing.T) {
+
+			expectedBuckets[i] = true
+			cleanProtocolMaps(t, "postgres", monitor.ebpfProgram.Manager.Manager)
+
+			require.NoError(t, monitor.Resume())
+			runQueryForBucket(t, pgClient, i*ebpf.MsgCountBucketSize)
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:953]: for i == 0, you do nothing

File: pkg/network/usm/postgres_monitor_test.go [hunk 4]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+	pg, err := postgres.NewPGXClient(postgres.ConnectionOptions{
+		ServerAddress: serverAddress,
+		EnableTLS:     isTLS,
+	})
+	require.NoError(t, err)
+	require.NoError(t, pg.Ping())
+	return pg
+}
+
+// createLargeTable It runs a postgres query to create a table large enough to retrieve long responses later.
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:972]:
```suggestion
// createLargeTable runs a postgres query to create a table large enough to retrieve long responses later.
```

File: pkg/network/usm/postgres_monitor_test.go [hunk 5]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+		return nil, fmt.Errorf("unable to get %q map: %s", mapName, err)
+	}
+	if err := mp.Lookup(unsafe.Pointer(&zero), unsafe.Pointer(pgKernelTelemetry)); err != nil {
+		return nil, fmt.Errorf("unable to lookup %q map: %s", mapName, err)
+	}
+	return pgKernelTelemetry, nil
+}
+
+// compareMessagesCount returns true if the expected bucket is non-empty
+// each buckets may have 1 or 2 hits
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:1023]: > each buckets may have 1 or 2 hits

what do you mean?
that might be relevant for a single test, but not true for the general case

File: pkg/network/usm/postgres_monitor_test.go [hunk 6]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+		found, err := getKernelTelemetry(monitor, tls)
+		require.NoError(t, err)
+		require.NotNil(t, found)
+		actual = found
+		return compareMessagesCount(found, expected)
+	}, time.Second*2, time.Millisecond*100)
+	if t.Failed() {
+		t.Logf("\nexpected telemetry:\n %+v;\nactual telemetry:\n %+v", expected, actual)
+	}
+}
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:1001]: Did you check this code during a failure?

Comment (guyarb) [reply pkg/network/usm/postgres_monitor_test.go:1001]: a.k.a, `getKernelTelemetry` fails, or we didn't find telemetry?

File: pkg/network/usm/postgres_monitor_test.go [hunk 7]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+	var actual *ebpf.PostgresKernelMsgCount
+	assert.Eventually(t, func() bool {
+		found, err := getKernelTelemetry(monitor, tls)
+		require.NoError(t, err)
+		require.NotNil(t, found)
+		actual = found
+		return compareMessagesCount(found, expected)
+	}, time.Second*2, time.Millisecond*100)
+	if t.Failed() {
+		t.Logf("\nexpected telemetry:\n %+v;\nactual telemetry:\n %+v", expected, actual)
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:999]:
```suggestion
		t.Logf("expected telemetry:\n %+v;\nactual telemetry:\n %+v", expected, actual)
```

File: pkg/network/usm/postgres_monitor_test.go [hunk 8]
```diff
@@ -889,3 +893,142 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+
+// runQueryForBucket The client sends a SELECT query to postgres with the specified SQL row limit of <N>.
+// For each SELECT request of <N> rows, the TCP response packet from Postgres is expected to contain
+//
+//	4 + <N> postgres messages in the format:
+//
+// "Bind completion", "Row description", "data row" 1...N, "command completion", "ready for query".
+func runQueryForBucket(t *testing.T, pg *postgres.PGXClient, limitCount int) {
+	require.NoError(t, pg.RunQuery(generateSelectLimitQuery(limitCount)))
+}
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:986]: I'm not sure how the documentation fits here
Also, I don't think a single that is being used only once, deserves a function

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 9]
```diff
@@ -210,11 +234,7 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
         return 0;
     }
 
-    if (iteration_value->iteration >= POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES) {
-        return 0;
-    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:215]: why did you remove this check?

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 10]
```diff
@@ -226,33 +246,47 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
...
         if (header.message_tag == POSTGRES_COMMAND_COMPLETE_MAGIC_BYTE) {
-            handle_command_complete(&conn_tuple, transaction);
             found_command_complete = true;
             break;
         }
         // We didn't find a command complete message, so we advance the data offset to the end of the message.
         // reminder, the message length includes the size of the payload, 4 bytes of the message length itself, but not
         // the message tag. So we need to add 1 to the message length to jump over the entire message.
         pktbuf_advance(pkt, header.message_len + 1);
+        iteration_value->msg_count++;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:262]: redundant operation. you can declare `message_count` outside from the loop, and call `iteration_value->msg_count += message_count` only once.
The current code is going to be executed `POSTGRES_MAX_MESSAGES_PER_TAIL_CALL` times, and generates redundant complexity and increases instruction count

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 11]
```diff
@@ -226,33 +246,47 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
...
     }
 
+    if (found_command_complete || (read_result == READ_FRAGMENTED)) {
+        // if completion was found or the packet was fragmented, then report the transaction and clear it from the map.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:266]: Handling command complete if we're dealing with a fragmented packet is wrong and must be reverted

It damages our reported latency per message and makes the metric unreliable

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 12]
```diff
@@ -226,33 +246,47 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
...
-    iteration_value->iteration += 1;
-    iteration_value->data_off = pktbuf_data_offset(pkt);
+    if ((read_result == READ_FRAGMENTED) || (iteration_value->msg_count == (PG_KERNEL_MAX_MESSAGES - 1))) {
+        // the packet was fragmented or max messages - add counter to the last bucket and stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->pg_messages_count_buckets[PG_KERNEL_MSG_COUNT_NUM_BUCKETS], 1);
 
-    // If the maximum number of tail calls has been reached, we can skip invoking the next tail call.
-    if (iteration_value->iteration >= POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES) {
         return 0;
     }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:282]: why did you remove this check?
What code piece promise we are executing up to `POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES` times?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 13]
```diff
@@ -29,18 +29,25 @@ static __always_inline void postgres_batch_enqueue_wrapper(conn_tuple_t *tuple,
...
+
+        return READ_OK;
     }
-    pktbuf_load_bytes(pkt, data_off, header, sizeof(struct pg_message_header));
-    // Converting the header to host byte order.
-    header->message_len = bpf_ntohl(header->message_len);
-    return true;
+    if (data_off == data_end) {
+        return READ_END;
+    }
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:48]: Typically a TCP packet contains many messages and only the last message reaches the end of the packet, so the data offset less than end of data condition is more common and should be checked first.

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 9 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:215]: Because I do not count iterations, I count total processed messages

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 12 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:282]: [Line 277 ](https://github.com/DataDog/datadog-agent/blob/aed60c21f5edccf132acc4b414a512231ce358d7/pkg/network/ebpf/c/protocols/postgres/decoding.h#L277) stops further iteration is reached allowed max messages

File: pkg/network/ebpf/c/protocols/postgres/decoding-maps.h [hunk 14]
```diff
@@ -15,4 +15,10 @@ BPF_PERCPU_ARRAY_MAP(postgres_scratch_buffer, postgres_event_t, 1)
 // Maintains the current state of tail calls for each Postgres message.
 BPF_PERCPU_ARRAY_MAP(postgres_iterations, postgres_tail_call_state_t, 1)
 
+/* Postgres telemetry maps in kernel provide empirical statistics on the number of processed Postgres messages.
+ * use only key 0, value is postgres telemetry object.
+ */
+BPF_ARRAY_MAP(postgres_plain_msg_count, postgres_kernel_msg_count_t, 1)
+BPF_ARRAY_MAP(postgres_tls_msg_count, postgres_kernel_msg_count_t, 1)
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding-maps.h]: changed code to use one map

File: pkg/network/ebpf/c/protocols/postgres/types.h [hunk 15]
```diff
@@ -7,10 +7,10 @@
 #define POSTGRES_BUFFER_SIZE 160
 
 // Represents the maximum number of tail calls we can use to process a single message.
-#define POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES 1
+#define POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES 5
 
 // Represents the maximum number of messages we process in a single tail call.
-#define POSTGRES_MAX_MESSAGES_PER_TAIL_CALL 80
+#define POSTGRES_MAX_MESSAGES_PER_TAIL_CALL 70
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/types.h]: I've decreased number of tail calls, now 3 tail calls, 60 messages each call

File: pkg/network/ebpf/c/protocols/postgres/types.h [hunk 16]
```diff
@@ -30,10 +30,41 @@ typedef struct {
 } postgres_event_t;
 
 typedef struct {
-    __u8 iteration;
+    __u32 msg_count;
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/types.h]: changed to use __u8

File: pkg/network/ebpf/c/protocols/postgres/types.h [hunk 17]
```diff
@@ -30,10 +30,41 @@ typedef struct {
...
+    // detected TCP packet fragmentation
+    READ_FRAGMENTED = 2,
+} postgres_read_result_t;
+
+// Postgres communication operates via a continuous message stream.
+// Gather empirical statistics on the number of messages processed by the program.
+// These statistics enable optimization of the monitoring code.
+// Collect counters for each subsequent bucket:
+// 0 to (BUCKET_SIZE-1), BUCKET_SIZE to (BUCKET_SIZE*2-1), (BUCKET_SIZE*2) to (BUCKET_SIZE*3-1), ...
+#define PG_KERNEL_MSG_COUNT_BUCKET_SIZE 35
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/types.h]: I've refactored how buckets are counted, first bucket is now 100 messages, then 4 buckets of 20 messages

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 11 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h]: removed condition of fragmented packet

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 12 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:291]: I've changed condition to stop iterator

File: pkg/network/usm/postgres_monitor_test.go (hunk 8 above)
Comment (yuri-lipnesh) [reply pkg/network/usm/postgres_monitor_test.go]: I replaced this function with direct call

File: pkg/network/protocols/postgres/protocol.go [hunk 18]
```diff
@@ -250,3 +271,44 @@ func (p *protocol) setupMapCleaner(mgr *manager.Manager) {
 
 	p.mapCleaner = mapCleaner
 }
+
+func (p *protocol) startKernelTelemetry(mgr *manager.Manager) {
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/protocol.go:280]: I've unified two maps

File: pkg/network/ebpf/c/protocols/postgres/types.h [hunk 19]
```diff
@@ -7,10 +7,13 @@
 #define POSTGRES_BUFFER_SIZE 160
 
 // Represents the maximum number of tail calls we can use to process a single message.
-#define POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES 1
+#define POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES 3
```
Comment (yuri-lipnesh) [inline pkg/network/ebpf/c/protocols/postgres/types.h:10]: this constant is no longer used separately, may be can remove it and use POSTGRES_MAX_TOTAL_MESSAGES if needed?

File: pkg/network/usm/postgres_monitor_test.go [hunk 20]
```diff
@@ -890,3 +892,164 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+		return nil, fmt.Errorf("unable to get %q map: %s", mapName, err)
+	}
+	if err := mp.Lookup(unsafe.Pointer(&key), unsafe.Pointer(pgKernelTelemetry)); err != nil {
+		return nil, fmt.Errorf("unable to lookup %q map: %s", mapName, err)
+	}
+	return pgKernelTelemetry, nil
+}
+
+// compareMessagesCount returns true if the expected bucket is non-empty
+func compareMessagesCount(found *ebpf.PostgresKernelMsgCount, expected [ebpf.MsgCountNumBuckets]bool) bool {
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:1030]: why do you pass `ebpf.PostgresKernelMsgCount`?

File: pkg/network/usm/postgres_monitor_test.go [hunk 21]
```diff
@@ -890,3 +892,164 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+// validateKernelExceedingMax check for exceeding the maximum number of buckets
+func validateKernelExceedingMax(t *testing.T, monitor *Monitor, tls bool) {
+	var actual *ebpf.PostgresKernelMsgCount
+	assert.Eventually(t, func() bool {
+		found, err := getKernelTelemetry(monitor, tls)
+		require.NoError(t, err)
+		require.NotNil(t, found)
+		actual = found
+		return found.Reached_max_messages > 0
+	}, time.Second*2, time.Millisecond*100)
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:1051]: you cannot use `require.*` inside `assert.Eventually`

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:48]: I believe you didn't follow my comment
if `data_off == data_end`, then for sure, the condition `data_off + sizeof(struct pg_message_header) <= data_end` will return false.
I suggested to switch the order of the if-clauses

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 22]
```diff
@@ -108,6 +115,24 @@ static int __always_inline skip_string(pktbuf_t pkt, int message_len) {
...
+            .map = &postgres_telemetry,
+            .key = (void*)&plain_key,
+        },
+        [PKTBUF_TLS] = {
+            .map = &postgres_telemetry,
+            .key = (void*)&tls_key,
+        },
+    };
+    return pktbuf_map_lookup(pkt, pg_telemetry_lookup_opt);
+}
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:134]: perfect

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 9 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:215]: I'm not following why you removed the check, it is still a valid check, you might just need to modify the condition

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 23]
```diff
@@ -240,19 +262,39 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
...
-    // We save the current data offset and increment the iteration counter.
-    iteration_value->iteration += 1;
-    iteration_value->data_off = pktbuf_data_offset(pkt);
 
-    // If the maximum number of tail calls has been reached, we can skip invoking the next tail call.
-    if (iteration_value->iteration >= POSTGRES_MAX_TAIL_CALLS_FOR_MAX_MESSAGES) {
+    if (iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)) {
+        // reached max messages, add counter and stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->reached_max_messages, 1);
+        return 0;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:284]: you can reach to both `POSTGRES_MAX_TOTAL_MESSAGES` and have a fragmented read.
I don't think the return here is beneficial

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:48]: It is right what you are saying, my point is that most frequently happens case when `(data_off + sizeof(struct pg_message_header) <= data_end)` is TRUE and `(data_off == data_end)` is FALSE. Only sometimes (at the end of packet) condition `(data_off == data_end) ` is TRUE and `(data_off + sizeof(struct pg_message_header) <= data_end)`  is FALSE. Checking frequent condition first gives performance advantage.

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 23 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:284]: with current implementation conditions `(iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)) `and `(read_result == READ_FRAGMENTED)` may not happen simultaneously, because fragmentation is detected at the beginning of TCP packet

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 24]
```diff
@@ -240,19 +262,39 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
...
     if (found_command_complete) {
+        handle_command_complete(&conn_tuple, transaction);
+
+        // save messages counter in bucket.
+        if (iteration_value->total_msg_count < PG_KERNEL_MSG_COUNT_FIRST_BUCKET) {
+            __sync_fetch_and_add(&pg_msg_counts->msg_count_buckets[0], 1);
+        } else {
+            __u8 bucket_idx = ((__u8)(iteration_value->total_msg_count - PG_KERNEL_MSG_COUNT_FIRST_BUCKET) / PG_KERNEL_MSG_COUNT_BUCKET_SIZE) + 1;
+            bucket_idx = (bucket_idx >= PG_KERNEL_MSG_COUNT_NUM_BUCKETS) ? (PG_KERNEL_MSG_COUNT_NUM_BUCKETS - 1) : bucket_idx;
+            __sync_fetch_and_add(&pg_msg_counts->msg_count_buckets[bucket_idx], 1);
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h]: I replaced 'if's with ternary operator like http2 does

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 25]
```diff
@@ -29,18 +29,25 @@ static __always_inline void postgres_batch_enqueue_wrapper(conn_tuple_t *tuple,
...
+        header->message_len = bpf_ntohl(header->message_len);
+
+        return READ_OK;
+    }
+    if (data_off == data_end) {
+        return READ_END;
+    }
+    // the offset is outside the packet boundaries
+    return READ_FRAGMENTED;
 }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:51]: Reporting `READ_OK`, `READ_END` and `READ_FRAGMENTED` does not bring any value to the PR.
Eventually, you're only checking if the result was `READ_OK` or not which is equivalent for the previous version of a boolean

Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:51]: I increase counter `pg_msg_counts->fragmented_packets` when `READ_FRAGMENTED` and stop iterating tail calls when `READ_END`

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:51]: you're using it only once, and we can get the same outcome without modifying that internal function.

File: pkg/network/usm/postgres_monitor_test.go [hunk 26]
```diff
@@ -890,3 +891,166 @@ func createFragment(fragment []byte) [ebpf.BufferSize]byte {
...
+	serverAddress := net.JoinHostPort(serverHost, postgresPort)
+	require.NoError(t, postgres.RunServer(t, serverHost, postgresPort, isTLS))
+	waitForPostgresServer(t, serverAddress, isTLS)
+
+	monitor := setupUSMTLSMonitor(t, getPostgresDefaultTestConfiguration(isTLS))
+	if isTLS {
+		utils.WaitForProgramsToBeTraced(t, consts.USMModuleName, GoTLSAttacherName, os.Getpid(), utils.ManualTracingFallbackEnabled)
+	}
+	pgClient := setupPGClient(t, serverAddress, isTLS)
+	defer func() {
```
Comment (guyarb) [inline pkg/network/usm/postgres_monitor_test.go:932]: use t.Cleanup

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 25 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:51]: I've restored previous `read_message_header()` function

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 23 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:288]: I've removed `return` under this condition

File: pkg/network/ebpf/c/protocols/postgres/types.h [hunk 27]
```diff
@@ -30,10 +33,36 @@ typedef struct {
...
 typedef struct {
-    __u8 iteration;
+    __u8 total_msg_count;
     // Saving the packet data offset is crucial for maintaining the current read position and ensuring proper utilization
     // of tail calls.
     __u32 data_off;
 } postgres_tail_call_state_t;
 
+typedef enum
+{
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/types.h:43]: I've removed this enum in recent commit

File: pkg/network/protocols/postgres/telemetry.go [hunk 28]
```diff
@@ -178,3 +179,44 @@ func (t *Telemetry) Log() {
...
+func newKernelTelemetry() *kernelTelemetry {
+	metricGroup := libtelemetry.NewMetricGroup("usm.postgres", libtelemetry.OptStatsd)
+	kernelTel := &kernelTelemetry{
+		metricGroup: metricGroup,
+	}
+	kernelTel.reachedMaxMessages = libtelemetry.NewTLSAwareCounter(metricGroup, "messages_count_max_messages")
+	kernelTel.fragmentedPackets = libtelemetry.NewTLSAwareCounter(metricGroup, "messages_count_fragmented_packets")
+
+	for i := range kernelTel.msgCountBuckets {
+		kernelTel.msgCountBuckets[i] = libtelemetry.NewTLSAwareCounter(metricGroup, "messages_count_bucket_"+(strconv.Itoa(i+1)))
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go:201]: fixed

File: pkg/network/protocols/postgres/telemetry.go [hunk 29]
```diff
@@ -178,3 +179,44 @@ func (t *Telemetry) Log() {
...
+	t.reachedMaxMessages.Set(int64(kernCounts.Reached_max_messages), isTLS)
+	t.fragmentedPackets.Set(int64(kernCounts.Fragmented_packets), isTLS)
+
+	for i := range t.msgCountBuckets {
+		v := kernCounts.Msg_count_buckets[i]
+		t.msgCountBuckets[i].Set(int64(v), isTLS)
+	}
+
+	s := t.metricGroup.Summary()
+	log.Debugf("postgres kernel telemetry, isTLS=%t, summary: %s", isTLS, s)
```
[the remark this answers is not in the document]
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go:221]: I've added `ShouldLog` condition

File: pkg/network/protocols/postgres/telemetry.go [hunk 30]
```diff
@@ -178,3 +179,45 @@ func (t *Telemetry) Log() {
...
+	t.fragmentedPackets.Set(int64(kernCounts.Fragmented_packets), isTLS)
+
+	for i := range t.msgCountBuckets {
+		v := kernCounts.Msg_count_buckets[i]
+		t.msgCountBuckets[i].Set(int64(v), isTLS)
+	}
+	if log.ShouldLog(seelog.DebugLvl) {
+		s := t.metricGroup.Summary()
+		log.Debugf("postgres kernel telemetry, isTLS=%t, summary: %s", isTLS, s)
+	}
```
Comment (guyarb) [inline pkg/network/protocols/postgres/telemetry.go:222]: viewing that again, why do you log the telemetry when we update it?
it is irrelevant to the update operation and should not be tied together

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 31]
```diff
@@ -240,20 +283,33 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
         // the message tag. So we need to add 1 to the message length to jump over the entire message.
         pktbuf_advance(pkt, header.message_len + 1);
     }
+    iteration_value->total_msg_count += messages_count;
 
     if (found_command_complete) {
+        handle_command_complete(&conn_tuple, transaction);
+        update_msg_count_telemetry(pg_msg_counts, iteration_value->total_msg_count);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:290]: what is the value of this PR if you update the telemetry only if we've found a command complete?

File: pkg/network/ebpf/c/protocols/postgres/decoding.h [hunk 32]
```diff
@@ -240,20 +283,33 @@ static __always_inline bool handle_response(pktbuf_t pkt, conn_tuple_t conn_tupl
...
+        // stop the iterator if the end of the TCP packet is reached.
+        return 0;
+    }
+    if (read_result == false) {
+        // the packet was fragmented, add counter stop iterating.
+        __sync_fetch_and_add(&pg_msg_counts->fragmented_packets, 1);
         return 0;
     }
 
+    // We didn't find a command complete message, so we need to continue processing the packet.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/postgres/decoding.h:309]: if `iteration_value->total_msg_count >= (POSTGRES_MAX_TOTAL_MESSAGES - 1)` you're performing additional tail call (that will bail out).
But each tail call has its performance impact, so if you don't need to trigger a new tail call - don't

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 31 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:290]: and what is a packet has only 80 messages in it, without having `command complete`?

File: pkg/network/protocols/postgres/telemetry.go [hunk 33]
```diff
@@ -216,8 +216,12 @@ func (t *kernelTelemetry) update(kernCounts *ebpf.PostgresKernelMsgCount, isTLS
 		v := kernCounts.Msg_count_buckets[i]
 		t.msgCountBuckets[i].Set(int64(v), isTLS)
 	}
+}
+
+// DebugLog logs summary of telemetry
+func (t *kernelTelemetry) DebugLog() {
```
Comment (guyarb) [inline pkg/network/protocols/postgres/telemetry.go:222]:
```suggestion
// Log the summary of telemetry
func (t *kernelTelemetry) Log() {
```

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 31 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:290]: The accuracy of the Postgres handler depends on its ability to detect Postgres completion commands. When reading a TCP packet, the following outcomes are possible:
1. A completion command is found.
2. The packet does not contain a completion command.
3. The packet is fragmented.
4. The maximum message counter is reached.
Cases 3 and 4 reduce accuracy, and their impact can be measured by the relative number of failed cases. In my initial implementation, I compared cases 3 and 4 against case 1, ignoring case 2 as irrelevant. However, based on my understanding of your previous comment, you disagreed with this approach. I have since updated the code to also count case 2.
To address your question: with the current implementation, a packet containing 80 messages without a completion command would be counted in the first bucket.
Ultimately, it involves a human interpretation of the results and an evaluation of the impact caused by the missed completion commands.

File: pkg/network/protocols/postgres/telemetry.go (hunk 33 above)
Comment (yuri-lipnesh) [reply pkg/network/protocols/postgres/telemetry.go:222]: done

File: pkg/network/ebpf/c/protocols/postgres/decoding.h (hunk 31 above)
Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:290]: it will be counted by field `pg_msg_counts->reached max_messages` as it exceeds allowed maximum. Whenever outcome is 3 or 4 we cannot say if completion command was in packet or was not.

Comment (yuri-lipnesh) [reply pkg/network/ebpf/c/protocols/postgres/decoding.h:290]: Addition. All possible outcomes handled on following lines. Line 290 adds to bucket when discovered completion command, Line 297 saves counter when reached maximum of messages, line 302 covers case of end of packet, line 307 adds to counter of broken packet.

**Review by guyarb — APPROVED** [review]
